### ISSUE TITLE
SC2034 false positive in heredoc

### ISSUE THREAD

**Issue by mvanorder** [koalaman/shellcheck#1806]
shellcheck version: 0.7.0

When a variable is only used in heredoc shellcheck doesn't see it as being used.

#### Here's a snippet or screenshot that shows the problem:

```sh
pub_key=$(cat "${HOME}/.ssh/id_rsa.pub")

ssh jumpserver << 'ENDSSH'
        [ ! -d .ssh ] && mkdir -m700 .ssh
        grep "${pub_key}" .ssh/authorized_keys > /dev/null || echo "${pub_key}" >> .ssh/authorized_keys
ENDSSH
```

#### Here's what shellcheck currently says:

In test line 42:
    pub_key=$(cat "${HOME}/.ssh/id_rsa.pub")
    ^-----^ SC2034: pub_key appears unused. Verify use (or export if used externally).

#### Here's what I wanted or expected to see:

Nothing

**Comment by gromgit** [comment]
You quoted the heredoc delimiter. From the `bash` man page:
> If any part of [the heredoc] `word` is quoted, the `delimiter` is the result of quote removal on `word`, and **the lines in the here-document are not expanded**.

Try replacing `ssh jumpserver` with `cat`. You'll see that `$pub_key` is not substituted in the heredoc, and since you don't deference `pub_key` anywhere else in your snippet, SC2034 is perfectly correct.

**Comment by mvanorder** [comment]
I see what's wrong.

While I was going through and quoting everything, I didn't read this message.

```
    ssh jumpserver << ENDSSH
                      ^----^ SC2087: Quote 'ENDSSH' to make here document expansions happen on the server side rather than on the client.
```

Quoting ENDSSH actually broke the script.
